server/src/main/java/org/opensearch/rest/pagination/IndexBasedPaginationStrategy.java

```diff
@@ -0,0 +1,191 @@
+        this.indicesFromRequestedToken = metadataListForRequestedToken.stream()
+            .map(metadata -> metadata.getIndex().getName())
+            .collect(Collectors.toList());
+        this.paginatedQueryResponse = getPaginatedResponseForRequestedToken(paginatedQueryRequest.getSize(), sortedIndicesList);
+    }
+
+    private static Predicate<IndexMetadata> getMetadataListFilter(String requestedTokenStr, String sortOrder) {
+        boolean isAscendingSort = sortOrder.equals(PAGINATED_QUERY_ASCENDING_SORT);
+        IndexStrategyToken requestedToken = Objects.isNull(requestedTokenStr) || requestedTokenStr.isEmpty()
+            ? null
+            : new IndexStrategyToken(requestedTokenStr);
+        if (Objects.isNull(requestedToken)) {
+            return indexMetadata -> true;
+        }
+        return indexMetadata -> {
+            if (indexMetadata.getIndex().getName().equals(requestedToken.nameOfLastRespondedIndex)) {
+                return false;
+            } else if (indexMetadata.getCreationDate() == requestedToken.creationTimeOfLastRespondedIndex) {
+                return isAscendingSort
+                    ? indexMetadata.getIndex().getName().compareTo(requestedToken.nameOfLastRespondedIndex) > 0
+                    : indexMetadata.getIndex().getName().compareTo(requestedToken.nameOfLastRespondedIndex) < 0;
+            }
+            return isAscendingSort
+                ? indexMetadata.getCreationDate() > requestedToken.creationTimeOfLastRespondedIndex
+                : indexMetadata.getCreationDate() < requestedToken.creationTimeOfLastRespondedIndex;
+        };
+    }
+
+    private static Comparator<IndexMetadata> getMetadataListComparator(String sortOrder) {
+        boolean isAscendingSort = sortOrder.equals(PAGINATED_QUERY_ASCENDING_SORT);
+        return (metadata1, metadata2) -> {
+            if (metadata1.getCreationDate() == metadata2.getCreationDate()) {
+                return isAscendingSort
+                    ? metadata1.getIndex().getName().compareTo(metadata2.getIndex().getName())
+                    : metadata2.getIndex().getName().compareTo(metadata1.getIndex().getName());
+            }
+            return isAscendingSort
+                ? Long.compare(metadata1.getCreationDate(), metadata2.getCreationDate())
+                : Long.compare(metadata2.getCreationDate(), metadata1.getCreationDate());
+        };
+    }
+
+    private List<IndexMetadata> getMetadataListForRequestedToken(
+        List<IndexMetadata> sortedIndicesList,
+        PaginatedQueryRequest paginatedQueryRequest
+    ) {
+        if (sortedIndicesList.isEmpty()) {
+            return new ArrayList<>();
+        }
+        return sortedIndicesList.subList(0, Math.min(paginatedQueryRequest.getSize(), sortedIndicesList.size()));
+    }
+
+    private PaginatedQueryResponse getPaginatedResponseForRequestedToken(int pageSize, List<IndexMetadata> sortedIndicesList) {
```

Comment:
getResponseForToken

Reply:
this method is not returning actual response?

Reply:
As discussed, this is just for returning the token. Renamed appropriately.